### ISSUE TITLE
GitHub actions on MacOS failing since OPENSSL 1.0.2 removal

### ISSUE THREAD

**Issue by ycrumeyrolle** [dotnet/core#5787]
Related to https://github.com/actions/virtual-environments/issues/2089

I have an issue since the 14th December on my dotnet unit test workflow running on MacOS :
`The active test run was aborted. Reason: Test host process crashed : No usable version of libssl was found`
This occurs on netcoreapp3.1 with SDK 3.1.404.

`Current runner version: '2.275.1'
Operating System
  Mac OS X
  10.15.7
  19H15
Virtual Environment
  Environment: macos-10.15
  Version: 20201212.1`

**Comment by scalablecory** [comment]
@safern can you help route this?

**Comment by safern** [comment]
This is happening in `dotnet test` so I would start with @nohwnd

**Comment by nohwnd** [comment]
@ycrumeyrolle  are you still facing that issue? I tried making a pipeline based on the info you shared and it works. https://github.com/nohwnd/macOSTestIssue/runs/1702266623?check_suite_focus=true I am not using any special version of test platform, not even the latest packages. 

If you are still seeing the issue, plese check if there is something significantly different from what you are doing.

**Comment by ycrumeyrolle** [comment]
I am trying to reproduce the issue with your sample application. Without success for the moment...

**Comment by ycrumeyrolle** [comment]
@vcsjones @bartonjs we are able to reproduce the issue: 
https://github.com/nohwnd/macOSTestIssue/runs/1742981935?check_suite_focus=true
It seems to be related to AES-GCM only. Other cryptography algorithms were tested successfully (AES, SHA256, RSA, ECDH). 
AES-GCM trigger an error `No usable version of libssl was found`. 
This error occured exactly at: https://github.com/actions/virtual-environments/issues/2089

**Comment by vcsjones** [comment]
> It seems to be related to AES-GCM only. Other cryptography algorithms were tested successfully (AES, SHA256, RSA, ECDH).

`AesGcm` and `AesCcm` rely on the presence of OpenSSL on macOS, which is not installed by default. Other algorithms like AES, SHA2, RSA, ECDSA, etc. all rely on SecurityTransforms / CommonCrypto, which is built in to macOS. That at least explains why `AesGcm` is singled out.

This looks similar to https://github.com/dotnet/runtime/issues/46071.

**Comment by ycrumeyrolle** [comment]
According to https://github.com/actions/virtual-environments/blob/macOS-10.15/20201212.1/images/macos/macos-10.15-Readme.md, OpenSsl is installed by default on the GitHub MacOS image.
When running the command `openssl version`: `OpenSSL 1.1.1i  8 Dec 2020`

Maybe because of the openssl location ? ((/usr/local/opt/openssl -> /usr/local/Cellar/openssl@1.1/1.1.1i))

**Comment by ycrumeyrolle** [comment]
Just tried to add the environment variable for overriding the openssl path : 
`    CLR_OPENSSL_VERSION_OVERRIDE: /usr/local/Cellar/openssl@1.1/1.1.1i`
or 
`    CLR_OPENSSL_VERSION_OVERRIDE: /usr/local/Cellar/openssl@1.1/1.1.1i/lib`
Without success.

**Comment by bartonjs** [comment]
`CLR_OPENSSL_VERSION_OVERRIDE` only allows for controlling the (for macOS) version infix.  e.g. `CLR_OPENSSL_VERSION_OVERRIDE=1.1.1i` would look for libssl.1.1.1.i.dylib in the default library path.

Instead the problem is likely that the dylibs aren't in any default search path, so something like `DYLD_FALLBACK_LIBRARY_PATH =$DYLD_FALLBACK_LIBRARY_PATH:/usr/local/opt/openssl@1.1/lib` would be what's required.  (Using the path under opt means that you don't need to change the path every time a new revision is installed)

**Comment by ycrumeyrolle** [comment]
The variable DYLD_FALLBACK_LIBRARY_PATH is set to  /usr/local/opt/openssl@1.1/lib (this folder contains the file libssl.1.1.dylib)
https://github.com/nohwnd/macOSTestIssue/runs/1758078565?check_suite_focus=true#step:9:5

The variable CLR_OPENSSL_VERSION_OVERRIDE is set to  1.1.1i
https://github.com/nohwnd/macOSTestIssue/runs/1758078565?check_suite_focus=true#step:9:6

But the test is still failing with the same error
https://github.com/nohwnd/macOSTestIssue/runs/1758078565?check_suite_focus=true#step:9:18

**Comment by carlossanlop** [comment]
@jeffschwMSFT is this a problem with the host not finding the DLLs? Any chance you can help route this?

**Comment by jeffschwMSFT** [comment]
@carlossanlop I would start with our crypto experts first.

cc @GrabYourPitchforks

**Comment by danmoseley** [comment]
Do any of the DYLD_PRINT environment variables help show what's happening?

https://stackoverflow.com/questions/51504439/what-environment-variables-control-dyld#51504440

**Comment by danmoseley** [comment]
Especially DYLD_PRINT_LIBRARIES

**Comment by MartyIX** [comment]
Does anybody know about a possible workaround until this is fixed?

**Comment by danmoseley** [comment]
@MartyIX I don't think we know enough about the problem. Did you try the workaround that @bartonjs suggested, which did not work for @ycrumeyrolle? Assuming so, could you please set the environment variables I mention above and get more info?
